lib/linux_kernel/processor.c

```diff
@@ -644,6 +642,99 @@ M0_INTERNAL m0_processor_nr_t m0_processor_id_get(void)
 {
 	return smp_processor_id();
 }
+#elif defined CONFIG_AARCH64    /*CONFIG_AARCH64*/
+
+/**
+   Convert bitmap from one format to another. Copy cpumask bitmap to m0_bitmap.
+
+   @param dest -> Processors bitmap for Motr programs.
+   @param src -> Processors bitmap used by Linux kernel.
+   @param bmpsz -> Size of cpumask bitmap (src)
+
+   @pre Assumes memory is alloacted for outbmp and it's initialized.
+   
+   @see lib/processor.h
+   @see lib/bitmap.h
+ */
+
+static void processors_bitmap_copy(struct m0_bitmap *dest,
```

Comment:
Are those functions inside the section of CONFIG_AARCH64 copied and pasted from the x86 part? if they are the same, I don't think it is a good practice, as any bug in the x86 part will be bugs here. I will suggest not to put those common functions and APIs under #if #elif. Keep them as common functions, only put those having different implementations inside the cpu specific sections.

Reply:
These processor specific  function are not used in cort-motr code anywhere so far. They are kept as reference for future implementation. Since the implementation of each corresponding function may  vary across  processor. They need to be maintained separately

Reply:
I see m0_processor_nr_max(), m0_processors_online() are used by other part of code. For those processor interface APIs, there are supposed to be general and be able to handle different types of processor. But this may require some change of the code. For now it is fine to do in your way.

Reply:
Good point @siningwuseagate!

I see the code of at least this specific function - `processors_bitmap_copy()` - is completely identical to x86_64 version. So please drop it from here and use the generic version. And please, verify all other functions and avoid the code duplication wherever possible, as @siningwuseagate suggested.

Reply:
Though the code  are similar, it can not considered generic enough now. because they may  diverge from implementation prospective later on depending upon the processor architecture. if later on any issue is observed , it can not be ascertained to specific function  there ( either because of x86 or aarch64 changes)

Reply:
(1) IMO, the processor interface APIs such as m0_processor_init()/fini(), m0_processor_nr_max(), m0_processors_online() etc are generic in term of their definitions. Taking m0_processor_init() as example, although the x86 and aarch64 implementations are different, the definition of the API itself is generic, one way to differentiate the implementations inside the API is

M0_INTERNAL int m0_processors_init()
{
	int rc;
	M0_PRE(!processor_init);
#ifdef CONFIG_X86_64
	rc = processor_x86cache_create();
#elif defined CONFIG_AARCH64
        rc = 0;
#endif
	processor_init = (rc == 0);
	return M0_RC(rc);
}

2. for processors_bitmap_copy(), I see it use a quite generic linux function cpumask_test_cpu(). The archtect specific implementation is done by linux, can you explain why you think for this function, it will be a different implementations for different processors?

Reply:
There are few occurrence , where code can be  optimized  keeping an eye to space constraint like in am embedded system. But here saving space at the stake of more  complexity is less desirable from the point view of code maintenance. it  is hard to determine the genericity of each function now however. So  instead of cluttering each function with lots of #if, #else macro, it  is more  appreciable to keep the processor interface separate from each other altogether .

Reply:
I agree that we should not duplicated the code here and there.
Only when an API has different implementation, we need to do it with "#ifdef ... #endif".

Reply:
I agree with your point for the case when there is a lot of different code. But here, it seems, it's only the matter of a single line difference in one or two functions? If so - there is no point to code-duplicate the whole file!

It's better and handy to keep the difference between platforms close to one another so that the developer can see and grasp them in one screen page, rather than go and jump and search in different places of the file and try to compare.

Reply:
Updated the code to avoid the code duplication